ansible/roles/vm_set/files/mux_simulator.py

```diff
@@ -116,370 +112,535 @@ def run_cmd(cmdline):
-                            "action": "output",
-                            "out_port": "muxy-vms17-8-0"
-                        }
-                    ]
-                }
+    def __init__(self, vm_set, port_index):
+        self.isvalid = True
+        self.lock = threading.Lock()
+
+        self.vm_set = vm_set
+
+        self.port_index = port_index
+        self.bridge = adaptive_name(MUX_BRIDGE_TEMPLATE, vm_set, port_index)
+
+        self._init_ports()
+
+        # If the mux does not have valid ports attahed, it is invalid
+        if not self._get_ports():
+            self.isvalid = False
+            return
+
+        # Initilize the flows configured on the mux bridge
+        self._init_flows()
+        self._get_flows()
+
+        self.flap_counter = 0
+
+    def debug(self, msg):
+        app.logger.debug('bridge={}, {}'.format(self.bridge, msg))
+
+    def info(self, msg):
+        app.logger.info('bridge={}, {}'.format(self.bridge, msg))
+
+    def error(self, msg):
+        app.logger.error('bridge={}, {}'.format(self.bridge, msg))
+
+    def _init_ports(self):
+        self.ports = {
+            NIC: None,
+            UPPER_TOR: None,
+            LOWER_TOR: None
+        }
+        self.sides = {}
+
+    def _get_ports(self):
+        """Use the 'ovs-vsctl list-ports' command to get the ports attached to the mux bridge.
+
+        Example bridge name: 'mbr-vms17-8-0'. In the bridge name, 'vms17-8' is vm_set. '0' is port_index.
+        Example output of 'ovs-ofctl list-ports' command:
+
+        azure@str2-acs-serv-17:~$ sudo ovs-vsctl list-ports mbr-vms17-8-0
+        enp59s0f1.3216
+        enp59s0f1.3272
+        muxy-vms17-8-0
+
+        Returns:
+            boolean: Return False if it is not a valid mux bridge
+        """
+        out = run_cmd('ovs-vsctl list-ports {}'.format(self.bridge))
+        out_lines = out.splitlines()
+        if len(out_lines) != 3:
+            self.error('unexpected ports, found ports:\n{}'.format(out))
+            return False
+
+        tor_ports = []
+        nic_port = None
+        for port in out_lines:
+            if port.startswith(MUX_NIC_PREFIX):
+                nic_port = port
+            else:
+                tor_ports.append(port)
+
+        if len(tor_ports) != 2:
+            self.error('Wrong ToR ports, found ports:\n{}'.format(out))
+            return False
+        if not nic_port:
+            self.error('No NIC port, found ports:\n{}'.format(out))
+            return False
+
+        # Sort the ToR ports. Assume the upper ToR is assigned with lower VLAN.
+        tor_ports.sort()
+        self.ports = {
+            NIC: nic_port,
+            UPPER_TOR: tor_ports[0],
+            LOWER_TOR: tor_ports[1]
+        }
+        self.sides = {
+            nic_port: NIC,
+            tor_ports[0]: UPPER_TOR,
+            tor_ports[1]: LOWER_TOR
+        }
+
+        return True
+
+    def _init_active_standby(self, active_side):
+        if active_side is None:
```

Comment:
The function name is a little confusing. As far as I understand, the function is used to save the active/standby side after updating, how about make the function name more specific?
And we always call this function along with updating ```self.flow```. How about include the update of ```self.flow``` in this function?

Reply:
Yes, the function name is a little bit confusing. Since we have 4 attributes directly related with active/standby: self.active_side, self.active_port, self.standby_side, self.standby_port, it is tedious to set each of them when active/standby toggled. This function is a helper function for setting these 4 attributes. Value of `self.flows` has no 1-1 mapping with active/standby state. The logic of this helper function could be complicated if we include updating of `self.flows` in it.

Reply:
Updated function name to `_active_standby_state_helper`

Reply:
Make sense to me. Thanks